proto/query.proto

```diff
@@ -348,6 +348,10 @@ message ExecuteOptions {
   // WorkloadName specifies the name of the workload as indicated in query directives. This is used for instrumentation
   // in metrics and tracing spans.
   string WorkloadName = 15;
+
+  // criticality specifies the criticality of the query, between 0 and 100. This is leveraged by the transaction
+  // throttler to determine whether, under resource contention, a query should or should not be throttled.
+  string criticality = 16;
```

Comment:
this should be an integer right?

Reply:
No; if it is not passed by in the query comments, the integer will end up being 0 in `vttablet` as if the specified criticality was `0` instead of not being specified - in which case it should use whatever is set in the default criticality flag.

Reply:
I am not sure that's enough/would work, but maybe I am misunderstanding something. Let's say that it is defined as integer. This is parsed in the vtgate from the query directives and there are two options of interest for what I see as a potential issue:

1. It is not present.
2. It is passed by the client as zero.

If it is not present, it is not sent to the vttablet. Eventually the message becomes a go struct and because the data is not in the underlying proto message, the attribute of the `query.ExecuteOptions` struct is set to zero.

If it is passed by the client as zero, it is set to zero in the message, sent to vttablet, and set to zero in the `query.ExecuteOptions` struct. 

`vttablet` has no way to determine whether the criticality zero because it was not passed in the query or because that's what the user set on the comments. In the first case, it should overwrite it with what is set in the corresponding flag (which exists on the vttablet, but not in vtgate). On the second one, it should leave it as it is.

When passed as a string, the empty string `""` is clearly distinguishable from the string with a zero `"0"`, so vttablet knows whether to use the default criticality or use the zero in the `ExecuteOptions` struct.

An alternative is to set the flag for default criticality a part of vtgate instead of vttablet and always send them in the execute options, but from a previous PR, I was under the impression that we didn't want to send values that were explicitly set.

Another option is to have vtgate pass a "special" integer value (e.g. -1) if criticality is not present in the query directives and replace that in the vttablet with the default criticality. I can do that if you prefer, though to me it seems functionally equivalent to passing the empty string and checking for that for potential overriding in the vttablet.

Reply:
@ejortegau you're right that this is more complex than I thought. My previous experience with similar problems was that `vtgate` or `vtctl` did have access to the existing config (e.g. via `topo`) and could therefore send a complete and correct config object to `vttablet`.

This, an the fact that `0` is a perfectly valid value (where other times, it's a clear indicator for "no value").

I completely understand the logic of using an empty string as an indicator for "no value". I think a negative integer value is a good option. The thing with a string is the risk for errors, and the conventional idiom of always checking for errors, then reporting them, etc., which complicates the code.

I personally prefer a negative integer value as an indicator. If it's a sensible and trivial change for you, great, and let's do that. If it only complicates matters, forget it; it's not a hill to die on.

Reply:
Hi, @shlomi-noach , thanks for your input.
I am fine changing it to `-1` if that is preferred. The only caveat here is that criticality information would then *always* be set and sent by vtgate to vttablet, either because the user specified, or because they didn't and it was therefore set to `-1`. From a [previous PR](https://github.com/vitessio/vitess/pull/12394#discussion_r1130578759) I was under the impression that we didn't want to set and send values that were not explicitly  set  in the directives by the client. If we use `""`, we can follow that. If we use `-1`, we need to explicitly set and send it, against the advice on that PR. 

I will do whichever you prefer.

Reply:
All right, let's be consistent with https://github.com/vitessio/vitess/pull/12394#discussion_r1130578759 then. Keep it a string.

Reply:
I am not still aligned why would anyone want to change the criticality default from 0 to any other value?
@ejortegau do you have a use case? Will your production deployment going to have a different value?

Reply:
@harshit-gangal in my understanding you will have higher priority operations (like your normal OLTP) and lower level operations, where you may choose a different priority. And in my understanding `0`, the default, is "highest priority".

Reply:
With the current implementation, 0 is the lowest.

The reason why we want to have the default to something different than 0 is onboarding. When onboarding, many/most/all of your queries will lack criticality information, and you don't want them to start being throttled. Hence the default we will is when onboarding will be 100 (highest, corresponding to never throttle). As we make progress, our client app queries will start sending their own, actual criticalities. Eventually, we will no longer need to fill in the missing criticality from the query with a high default one, and instead can move it to be the lowest (the default of 0). This is the use case we have for it.

Reply:
@ejortegau I'm just trying to think about how this will behave for other Vitess users who are unaware of your specific intentions. For those users, if they ever activate tx throttler, their queries will have criticality `0`, meaning "always throttled" -- did I understand corrcetly?

Reply:
that is the default behaviour today with throttler enabled.

Reply:
Ah, right. That actually made sense, I was confused earlier.